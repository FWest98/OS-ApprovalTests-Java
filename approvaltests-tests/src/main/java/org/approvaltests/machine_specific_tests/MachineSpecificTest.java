package org.approvaltests.machine_specific_tests;

import com.spun.util.SystemUtils;
import com.spun.util.logger.SimpleLogger;
import org.junit.jupiter.api.BeforeEach;

import static org.junit.jupiter.api.Assumptions.assumeTrue;

public class MachineSpecificTest
{
  private static boolean DISPLAYED = false;
  @BeforeEach
  public void beforeMethod()
  {
    if (!MachineSpecific.isMachineConfiguredForTesting())
    {
      displayMessage();
      assumeTrue(false);
    }
  }
  private void displayMessage()
  {
    if (!DISPLAYED)
    {
      DISPLAYED = true;
      String message = String.format(
          "This machine isn't configured to run machine_specific_tests.\n" + "To run these either\n"
              + "  1) Set machine_specific_tests.MachineSpecific.FORCE_RUN=true\n"
              + "  2) Add \"%s\" to machine_specific_tests.MachineSpecific.MACHINES",
          SystemUtils.getComputerName());
      SimpleLogger.warning(message);
    }
  }
}
