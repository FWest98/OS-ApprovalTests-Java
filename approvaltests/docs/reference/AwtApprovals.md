<a id="top"></a>

# AwtApprovals

toc

## Paintables

### Why

Java AWT is a nice utility for rendering simple graphics, but it does not play well with CI systems,
especially when it is headless. Paintables provides an abstraction that allows you to use Awt
independently of OS and CI.

## How To

### create an animated gif

Sometimes you want to verify a series of steps. A great way to visualize this is with an animated
gif.

![Hexagonal Game of Life](https://github.com/isidore/HexGameOfLife/blob/master/src/test/java/org/gameoflife/hex/GameOfLifePanelTest.testCompellingSequence.approved.gif)

To create such a gif you pass the number of frames you want plus a function that returns a Paintable
for every frame. Here is an example for a simple expanding box:

snippet: SequencePaintables

**Note**: Method overloads allow specifying the time between frames or the time for each frame.



