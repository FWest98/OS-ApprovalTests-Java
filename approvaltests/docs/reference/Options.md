<a id="top"></a>

# Options

<!-- toc -->
## Contents

* [Introduction](#introduction)
* [Fluent Interface](#fluent-interface)
* [Reporters](#reporters)
* [Scrubbers](#scrubbers)
* [File Options](#file-options)
    * [File Extensions](#file-extensions)
* [Defaults](#defaults)
* [Adding to Existing Options object](#adding-to-existing-options-object)<!-- endToc -->

## Introduction

There are many things you might want to tweak with Approval Tests. `Options` is the entry-point for many of the changes.
It is on all `verify()` methods, as an optional parameter.

## Fluent Interface

`Options` utilizes a fluent interface, allowing you to chain together commands. Each returned object is a new copy.

<!-- snippet: specify_all_the_options -->
<a id='snippet-specify_all_the_options'></a>
```cpp
ApprovalTests::Options()
    .withReporter(ApprovalTests::QuietReporter())
    .withScrubber(ApprovalTests::Scrubbers::scrubGuid)
    .fileOptions().withFileExtension(".json")
```
<sup><a href='/tests/DocTest_Tests/DocTestApprovalTestTests.cpp#L50-L55' title='Snippet source file'>snippet source</a> | <a href='#snippet-specify_all_the_options' title='Start of snippet'>anchor</a></sup>
<!-- endSnippet -->

## Reporters

[Reporters](/doc/Reporters.md#top) launch diff tools upon failure.  Read how to configure them with `Options` [here](../Reporters.md#via-options):

## Scrubbers

[Scrubbers](/doc/explanations/Scrubbers.md#top) clean output to help remove inconsistent pieces of text, such as dates.  Read how to set up Scrubbers [here](../Scrubbers.md#configuring-scrubbers):
There are two ways to set a Scrubber.

## File Options

The `Options::FileOptions` class exists to customise the `.approved` and `.received` files in various ways.

For now, it just controls the file extension.

### File Extensions

If you want to change the file extension of both the approved and received files, use `withFileExtension()`.

<!-- snippet: basic_approval_with_file_extension -->
<a id='snippet-basic_approval_with_file_extension'></a>
```cpp
using namespace ApprovalTests;

Approvals::verify("text to be verified",
                  Options().fileOptions().withFileExtension(".xyz"));
```
<sup><a href='/tests/DocTest_Tests/core/OptionsExamples.cpp#L7-L12' title='Snippet source file'>snippet source</a> | <a href='#snippet-basic_approval_with_file_extension' title='Start of snippet'>anchor</a></sup>
<!-- endSnippet -->

**Note:** `withFileExtension()` returns an `Options` object, so it's possible to keep appending more `with...()` calls.

## Defaults

The default constructor for Options does:

* no scrubbing
* uses file extension `.txt`
* uses whatever is currently set as [the default reporter](#registering-a-default-reporter).

## Adding to Existing Options object

Because of the fluent options, you can modify a specific option from an existing Options object,
while retaining all other settings, and not changing the original object.

```cpp
verifyWithQuietReporter(std::string text, const Options& o)
{
    Approvals::verify(text, o.withReporter(QuietReporter());
}
```

---

[Back to User Guide](/doc/README.md#top)