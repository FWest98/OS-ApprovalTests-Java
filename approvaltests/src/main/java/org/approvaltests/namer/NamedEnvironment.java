package org.approvaltests.namer;

import java.util.Arrays;

import com.spun.util.logger.SimpleLogger;

public class NamedEnvironment implements AutoCloseable
{
  public NamedEnvironment(String info)
  {
    NamerFactory.additionalInformation = info;
  }
  @Override
  public void close()
  {
    NamerFactory.getAndClearAdditionalInformation();
  }
  public boolean isCurrentEnvironmentValidFor(String... environment)
  {
    if (Arrays.asList(environment).contains(NamerFactory.additionalInformation))
    {
      return true;
    }
    else
    {
      SimpleLogger
          .message(String.format("Not valid for current environment: \"%s\"", NamerFactory.additionalInformation));
      return false;
    }
  }
}
