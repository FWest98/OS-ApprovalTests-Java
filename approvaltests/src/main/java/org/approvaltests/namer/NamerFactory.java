package org.approvaltests.namer;

import org.lambda.functions.Function0;
import org.packagesettings.PackageLevelSettings;

import com.spun.util.StringUtils;
import com.spun.util.SystemUtils;

public class NamerFactory
{
  public static String additionalInformation;
  public static String getAndClearAdditionalInformation()
  {
    if (additionalInformation == null)
    { return ""; }
    String out = "." + additionalInformation;
    additionalInformation = null;
    return out;
  }
  public static ApprovalResults ApprovalResults = new ApprovalResults();
  public static NamedEnvironment asMachineSpecificTest(Function0<String> environmentLabeller)
  {
    return asMachineSpecificTest(environmentLabeller.call());
  }
  public static NamedEnvironment asMachineNameSpecificTest()
  {
    return asMachineSpecificTest(SystemUtils.getComputerName());
  }
  public static NamedEnvironment asOsSpecificTest()
  {
    return asMachineSpecificTest(new OsEnvironmentLabeller());
  }
  public static NamedEnvironment asMachineSpecificTest(String environmentName)
  {
    return new NamedEnvironment(environmentName);
  }
  public static MultipleFilesLabeller useMultipleFiles()
  {
    return new MultipleFilesLabeller();
  }
  public static String getSubdirectory()
  {
    return (String) PackageLevelSettings.getValueFor("UseApprovalSubdirectory");
  }
  public static String getApprovalBaseDirectory()
  {
    return (String) PackageLevelSettings.getValueFor("ApprovalBaseDirectory");
  }
  public static NamedEnvironment withParameters(Object... parameters)
  {
    return new NamedEnvironment(StringUtils.join(parameters, ".", o -> "" + o));
  }
}
