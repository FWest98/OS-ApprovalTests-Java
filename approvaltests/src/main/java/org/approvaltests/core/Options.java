package org.approvaltests.core;

import com.spun.util.ArrayUtils;
import org.approvaltests.Approvals;
import org.approvaltests.ReporterFactory;
import org.approvaltests.namer.ApprovalNamer;
import org.approvaltests.namer.NamerWrapper;
import org.approvaltests.scrubbers.NoOpScrubber;
import org.approvaltests.writers.ApprovalWriterFactory;
import org.approvaltests.writers.DefaultApprovalWriterFactory;

import java.util.HashMap;
import java.util.Map;

public class Options
{
  private enum Fields {
                       SCRUBBER, REPORTER, FILE_OPTIONS_FILE_EXTENSION, FILE_OPTIONS_NAMER, WRITER;
  }
  private final Map<Fields, Object> fields = new HashMap<>();
  public Options()
  {
  }
  public Options(Scrubber scrubber)
  {
    fields.put(Fields.SCRUBBER, scrubber);
  }
  public Options(ApprovalFailureReporter reporter)
  {
    fields.put(Fields.REPORTER, reporter);
  }
  private Options(Map<Fields, Object> fields, Fields key, Object value)
  {
    this.fields.putAll(fields);
    this.fields.put(key, value);
  }
  public ApprovalFailureReporter getReporter()
  {
    return ArrayUtils.getOrElse(fields, Fields.REPORTER, ReporterFactory::get);
  }
  public Options withReporter(ApprovalFailureReporter reporter)
  {
    return new Options(fields, Fields.REPORTER, reporter);
  }
  public Options withScrubber(Scrubber scrubber)
  {
    return new Options(fields, Fields.SCRUBBER, scrubber);
  }
  public String scrub(String input)
  {
    return getScrubber().scrub(input);
  }
  private Scrubber getScrubber()
  {
    return ArrayUtils.getOrElse(fields, Fields.SCRUBBER, () -> NoOpScrubber.INSTANCE);
  }
  public ApprovalWriter createWriter(Object o)
  {
    ApprovalWriterFactory factory = ArrayUtils.getOrElse(fields, Fields.WRITER,
        DefaultApprovalWriterFactory::getDefaultFactory);
    return factory.create(o, this);
  }
  public Options withWriter(ApprovalWriterFactory approvalWriterFactory)
  {
    return new Options(fields, Fields.WRITER, approvalWriterFactory);
  }
  public FileOptions forFile()
  {
    return new FileOptions(this.fields);
  }
  public static class FileOptions
  {
    private final Map<Fields, Object> fields;
    public FileOptions(Map<Fields, Object> fields)
    {
      this.fields = fields;
    }
    public Options withExtension(String fileExtensionWithDot)
    {
      if (!fileExtensionWithDot.startsWith("."))
      {
        fileExtensionWithDot = "." + fileExtensionWithDot;
      }
      return new Options(fields, Fields.FILE_OPTIONS_FILE_EXTENSION, fileExtensionWithDot);
    }
    public ApprovalNamer getNamer()
    {
      return ArrayUtils.getOrElse(fields, Fields.FILE_OPTIONS_NAMER, Approvals::createApprovalNamer);
    }
    public String getFileExtension()
    {
      return ArrayUtils.getOrElse(fields, Fields.FILE_OPTIONS_FILE_EXTENSION, () -> ".txt");
    }
    public Options withBaseName(String fileBaseName)
    {
      NamerWrapper approvalNamer = new NamerWrapper(() -> fileBaseName, getNamer());
      return new Options(fields, Fields.FILE_OPTIONS_NAMER, approvalNamer);
    }
    public Options withName(String fileBaseName, String extension)
    {
      NamerWrapper approvalNamer = new NamerWrapper(() -> fileBaseName, getNamer());
      HashMap<Fields, Object> newFields = new HashMap<>(fields);
      newFields.put(Fields.FILE_OPTIONS_FILE_EXTENSION, extension);
      return new Options(newFields, Fields.FILE_OPTIONS_NAMER, approvalNamer);
    }
  }
}
