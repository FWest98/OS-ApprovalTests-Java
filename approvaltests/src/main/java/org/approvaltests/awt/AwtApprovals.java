package org.approvaltests.awt;

import java.awt.Component;
import java.awt.Image;
import java.awt.image.BufferedImage;
import java.time.Duration;

import org.approvaltests.Approvals;
import org.approvaltests.core.Options;
import org.approvaltests.namer.NamedEnvironment;
import org.approvaltests.namer.NamerFactory;
import org.approvaltests.writers.ComponentApprovalWriter;
import org.approvaltests.writers.ImageApprovalWriter;
import org.approvaltests.writers.PaintableApprovalWriter;
import org.lambda.functions.Function1;

import com.spun.swing.Paintable;
import com.spun.util.Tuple;
import com.spun.util.images.ImageWriter;

public class AwtApprovals
{
  public static void verify(Image image)
  {
    verify(image, new Options());
  }
  public static void verify(Image image, Options options)
  {
    verify(ImageWriter.toBufferedImage(image), options);
  }
  public static void verify(BufferedImage bufferedImage)
  {
    Approvals.verify(bufferedImage, new Options());
  }
  public static void verify(BufferedImage bufferedImage, Options options)
  {
    Approvals.verify(new ImageApprovalWriter(bufferedImage), options);
  }
  public static void verify(Component c)
  {
    verify(c, new Options());
  }
  public static void verify(Component c, Options options)
  {
    try (NamedEnvironment env = NamerFactory.asOsSpecificTest())
    {
      Approvals.verify(new ComponentApprovalWriter(c), options);
    }
  }
  public static void verify(Paintable c)
  {
    verify(c, new Options());
  }
  public static void verify(Paintable c, Options options)
  {
    Approvals.verify(new PaintableApprovalWriter(c), options);
  }
  public static void verifySequence(int numberOfFrames, Function1<Integer, Paintable> sequenceRenderer)
  {
    verifySequence(numberOfFrames, sequenceRenderer, new Options());
  }
  public static void verifySequenceWithTimings(int numberOfFrames,
      Function1<Integer, Tuple<Paintable, Duration>> sequenceRenderer)
  {
    verifySequenceWithTimings(numberOfFrames, sequenceRenderer, new Options());
  }
  public static void verifySequenceWithTimings(int numberOfFrames,
      Function1<Integer, Tuple<Paintable, Duration>> sequenceRenderer, Options options)
  {
    Approvals.verify(new PaintableMultiframeWriter(numberOfFrames, sequenceRenderer), options);
  }
  public static void verifySequence(int numberOfFrames, Function1<Integer, Paintable> sequenceRenderer,
      Options options)
  {
    Approvals.verify(new PaintableMultiframeWriter(numberOfFrames,
        c -> new Tuple<>(sequenceRenderer.call(c), Duration.ofMillis(500))), options);
  }
}
