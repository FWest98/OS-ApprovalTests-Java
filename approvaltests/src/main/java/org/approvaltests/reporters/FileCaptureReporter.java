package org.approvaltests.reporters;

import java.io.IOException;

import com.spun.util.StringUtils;
import com.spun.util.io.FileUtils;
import com.spun.util.logger.SimpleLogger;
import org.approvaltests.core.ApprovalFailureReporter;

import com.spun.util.ObjectUtils;

public class FileCaptureReporter implements ApprovalFailureReporter
{
  private String message = "*** adding received file via FileCaptureReporter for further inspection";
  public FileCaptureReporter()
  {
  }
  public FileCaptureReporter(String message)
  {
    this.message = message;
  }
  @Override
  public void report(String received, String approved)
  {
    run("git add --force " + received);
    run(String.format("git commit -m '%s'", message));
    run("git push");
  }
  private void run(String command)
  {
    SimpleLogger.event(command);
    try
    {
      Process process = Runtime.getRuntime().exec(command);
      process.waitFor();
      final int exitValue = process.exitValue();
      if (exitValue != 0)
      {
        final String stdout = FileUtils.readStream(process.getInputStream());
        final String stderr = FileUtils.readStream(process.getErrorStream());
        SimpleLogger.warning(String.format("stdout:\n%s\nstderr:\n%s", stdout, stderr));
      }
    }
    catch (IOException | InterruptedException e)
    {
      throw ObjectUtils.throwAsError(e);
    }
  }
}
