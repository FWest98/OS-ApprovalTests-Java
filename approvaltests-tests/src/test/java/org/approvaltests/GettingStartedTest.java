package org.approvaltests;

import org.approvaltests.awt.AwtApprovals;
import org.approvaltests.core.Options;
import org.approvaltests.machine_specific_tests.TvGuide;
import org.approvaltests.namer.NamedEnvironment;
import org.approvaltests.namer.NamerFactory;
import org.approvaltests.reporters.FileCaptureReporter;
import org.approvaltests.reporters.UseReporter;
import org.approvaltests.reporters.windows.BeyondCompareReporter;
import org.junit.jupiter.api.Test;

import java.awt.*;

public class GettingStartedTest {
    @Test
    void testBasicCall() {
        String objectToBeVerified = "Now is the time for all good coders to come to the aid of their countrypersons";
        // begin-snippet: basic_verified_call
        Approvals.verify(objectToBeVerified);
        // end-snippet
    }
    
    // begin-snippet: verifying_strings
    @Test
    public void testBuildString() {
        /* Do */
        // create a string with "Approval" and append "Tests" to it
        String s = "Approval";
        s += " Tests";

        /* Verify */
        // Verify the resulting string
        Approvals.verify(s);
    }
    // end-snippet

    // begin-snippet: verifying_objects
    @Test
    public void testObject()
    {
        /* Do */
        // create an 100 x 200 rectangle with the top corner at (5, 10)
        Rectangle objectUnderTest = new Rectangle(5, 10, 100, 200);
        /* Verify */
        // Verify the rectangle is properly defined
        Approvals.verify(objectUnderTest.toString());
    }
    // end-snippet

    @Test
    public void testObjectWithJson()
    {
        /* Do */
        // create an 100 x 200 rectangle with the top corner at (5, 10)
        Rectangle objectUnderTest = new Rectangle(5, 10, 100, 200);
        /* Verify */
        // Verify the rectangle is properly defined
        // begin-snippet: verifying_objects_with_json
        JsonApprovals.verifyAsJson(objectUnderTest);
        // end-snippet
    }

    // begin-snippet: verifying_arrays
    @Test
    public void testArray() {
        /* Do */
        // create a String Array and set values in the indexes
        String[] s = new String[2];
        s[0] = "Approval";
        s[1] = "Tests";

        /* Verify */
        // Verify the array
        Approvals.verifyAll("Text", s);
    }
    // end-snippet

    @Test
    @UseReporter(FileCaptureReporter.class)
    public void testTvGuide() {
        try (NamedEnvironment namedEnvironment = NamerFactory.asOsSpecificTest()) {
            if (!namedEnvironment.isCurrentEnvironmentValidFor("Window_Server_2022")) {
                return;
            }
        }
        // begin-snippet: verifying_gui
        /* Do */
        // create a TV Guide and select a show for 3pm
        TvGuide tv = new TvGuide();
        tv.selectTime("3pm");

        /* Verify */
        // Verify the TvGuide
        AwtApprovals.verify(tv);
        // end-snippet
    }
}
