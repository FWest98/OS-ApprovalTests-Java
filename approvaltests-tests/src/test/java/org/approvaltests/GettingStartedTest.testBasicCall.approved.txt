Now is the time for all good coders to come to the aid of their countrypersons